presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java

```diff
@@ -86,441 +94,709 @@
+    {
+        checkArgument(sampleSize > 0, "sampleSize is expected to be greater than zero");
+
+        if (partitions.size() <= sampleSize) {
+            return partitions;
+        }
+
+        List<HivePartition> result = new ArrayList<>();
+
+        int samplesLeft = sampleSize;
+
+        HivePartition min = partitions.get(0);
+        HivePartition max = partitions.get(0);
+        for (HivePartition partition : partitions) {
+            if (partition.getPartitionId().compareTo(min.getPartitionId()) < 0) {
+                min = partition;
             }
-            columnStatistics.setDataSize(dataSize);
+            else if (partition.getPartitionId().compareTo(max.getPartitionId()) > 0) {
+                max = partition;
+            }
+        }
 
-            columnStatistics.setNullsFraction(nullsFraction);
-            tableStatistics.setColumnStatistics(hiveColumnHandle, columnStatistics.build());
+        result.add(min);
+        samplesLeft--;
+        if (samplesLeft > 0) {
+            result.add(max);
+            samplesLeft--;
         }
-        return tableStatistics.build();
+
+        if (samplesLeft > 0) {
+            HashFunction hashFunction = murmur3_128();
+            Comparator<Map.Entry<HivePartition, Long>> hashComparator = Comparator
+                    .<Map.Entry<HivePartition, Long>, Long>comparing(Map.Entry::getValue)
+                    .thenComparing(entry -> entry.getKey().getPartitionId());
+            partitions.stream()
+                    .filter(partition -> !result.contains(partition))
+                    .map(partition -> immutableEntry(partition, hashFunction.hashUnencodedChars(partition.getPartitionId()).asLong()))
+                    .sorted(hashComparator)
+                    .limit(samplesLeft)
+                    .forEachOrdered(entry -> result.add(entry.getKey()));
+        }
+
+        return unmodifiableList(result);
     }
 
-    private boolean isLowHighSupportedForType(Type type)
+    @VisibleForTesting
+    static void validatePartitionStatistics(SchemaTableName table, Map<String, PartitionStatistics> partitionStatistics)
     {
-        if (type instanceof DecimalType) {
-            return true;
-        }
-        if (type.equals(TINYINT)
-                || type.equals(SMALLINT)
-                || type.equals(INTEGER)
-                || type.equals(BIGINT)
-                || type.equals(REAL)
-                || type.equals(DOUBLE)
-                || type.equals(DATE)) {
-            return true;
-        }
-        return false;
+        partitionStatistics.forEach((partition, statistics) -> {
+            HiveBasicStatistics basicStatistics = statistics.getBasicStatistics();
+            OptionalLong rowCount = basicStatistics.getRowCount();
+            rowCount.ifPresent(count -> checkStatistics(count >= 0, table, partition, "rowCount must be greater than or equal to zero: %s", count));
```

Comment:
Would you consider checking that if one of these is zero, then the rest are zero as well? E.g. to catch the case when rowCount = 0, but fileCount and/or data size is > 0?

Reply:
I was going to do this as a separate PR